File: 98. Validate Binary Search Tree/98. Validate Binary Search Tree.dart
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *   int val;
 *   TreeNode? left;
 *   TreeNode? right;
 *   TreeNode([this.val = 0, this.left, this.right]);
 * }
 */
class Solution {
  bool isValidBST(TreeNode? root) {
    return _isValidBSTHelper(root, null, null);
  }
  
  bool _isValidBSTHelper(TreeNode? node, int? min, int? max) {
    if (node == null) return true;
    
    if ((min != null && node.val <= min) || (max != null && node.val >= max)) {
      return false;
    }
    
    return _isValidBSTHelper(node.left, min, node.val) &&
           _isValidBSTHelper(node.right, node.val, max);
  }
}

File: 98. Validate Binary Search Tree/98. Validate Binary Search Tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    bool isValidBST(TreeNode* root) {
        return isValidBSTHelper(root, LLONG_MIN, LLONG_MAX);
    }
    
private:
    bool isValidBSTHelper(TreeNode* node, long long min, long long max) {
        if (node == nullptr) return true;
        
        if (node->val <= min || node->val >= max) return false;
        
        return isValidBSTHelper(node->left, min, node->val) &&
               isValidBSTHelper(node->right, node->val, max);
    }
};